File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-11/29-Jump-Game-iii.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        /**
        * T: O(n) and S: O(n)
        */
        Set<Integer> visited = new HashSet<>();
        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.push(start);
        
        while (!q.isEmpty()) {
            int currIndex = q.removeFirst();
            if (arr[currIndex] == 0) {
                return true;
            }
            int leftIndex = currIndex - arr[currIndex];
            int rightIndex = currIndex + arr[currIndex];
            
            if (leftIndex >= 0 && !visited.contains(leftIndex)) {
                visited.add(leftIndex);
                q.push(leftIndex);
            }
            if (rightIndex < arr.length && !visited.contains(rightIndex)) {
                visited.add(rightIndex);
                q.push(rightIndex);
            }
        }
        return false;
    }
}

File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-11/29-Jump-Game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        '''
        T: O(n) and S: O(n)
        '''
        visited = set()
        q = []
        q.append(start)
        
        while q:
            currIndex = q.pop()
            if arr[currIndex] == 0:
                return True
            
            leftIndex = currIndex - arr[currIndex]
            rightIndex = currIndex + arr[currIndex]
            
            if leftIndex >= 0 and leftIndex not in visited:
                visited.add(leftIndex)
                q.append(leftIndex)
            
            if rightIndex < len(arr) and rightIndex not in visited:
                visited.add(rightIndex)
                q.append(rightIndex)
        
        return False
